File: queries/h10.py
'Returned Item Reporting Query (Q10)'

import ibis
from utils import add_date


def tpc_h10(con, DATE='1993-10-01'):
    customer = con.table('customer')
    orders = con.table('orders')
    lineitem = con.table('lineitem')
    nation = con.table('nation')

    q = customer
    q = q.join(orders, customer.C_CUSTKEY == orders.O_CUSTKEY)
    q = q.join(lineitem, lineitem.L_ORDERKEY == orders.O_ORDERKEY)
    q = q.join(nation, customer.C_NATIONKEY == nation.N_NATIONKEY)
    q = q.materialize()

    q = q.filter([
        (q.O_ORDERDATE >= DATE) & (q.O_ORDERDATE < add_date(DATE, dm=3)),
        q.L_RETURNFLAG == 'R',
    ])

    gq = q.group_by([
            q.C_CUSTKEY,
            q.C_NAME,
            q.C_ACCTBAL,
            q.C_PHONE,
            q.N_NAME,
            q.C_ADDRESS,
            q.C_COMMENT
    ])
    q = gq.aggregate(revenue=(q.L_EXTENDEDPRICE*(1-q.L_DISCOUNT)).sum())

    q = q.sort_by(ibis.desc(q.revenue))
    return q.limit(20)

File: queries/h10.R
tpc_h10 <- function(input_func, collect_func = dplyr::collect) {
  l <- input_func("lineitem") %>%
    select(l_orderkey, l_returnflag, l_extendedprice, l_discount) %>%
    filter(l_returnflag == "R") %>%
    select(l_orderkey, l_extendedprice, l_discount)

  o <- input_func("orders") %>%
    select(o_orderkey, o_custkey, o_orderdate) %>%
    # kludge, filter(o_orderdate >= "1993-10-01", o_orderdate < "1994-01-01") %>%
    filter(o_orderdate >= as.Date("1993-10-01"), o_orderdate < as.Date("1994-01-01")) %>%
    select(o_orderkey, o_custkey)

  lo <- inner_join(l, o,
                   by = c("l_orderkey" = "o_orderkey")) %>%
    select(l_extendedprice, l_discount, o_custkey)
  # first aggregate, then join with customer/nation,
  # otherwise we need to aggr over lots of cols

  lo_aggr <- lo %>% mutate(volume=l_extendedprice * (1 - l_discount)) %>%
    group_by(o_custkey) %>%
    summarise(revenue = sum(volume))

  c <- input_func("customer") %>%
    select(c_custkey, c_nationkey, c_name, c_acctbal, c_phone, c_address, c_comment)

  loc <- inner_join(c, lo_aggr, by = c("c_custkey" = "o_custkey"))

  locn <- inner_join(loc, input_func("nation") %>% select(n_nationkey, n_name),
                     by = c("c_nationkey" = "n_nationkey"))

  locn %>%
    select(c_custkey, c_name, revenue, c_acctbal, n_name,
           c_address, c_phone, c_comment) %>%
    arrange(desc(revenue)) %>%
    head(20) %>%
    collect_func()
}
